clients/src/main/java/org/apache/kafka/common/utils/ThroughputThrottler.java

```diff
@@ -96,12 +101,19 @@ public void throttle() {
 
         // If enough sleep deficit has accumulated, sleep a little
         if (sleepDeficitNs >= MIN_SLEEP_NS) {
-            long sleepMs = sleepDeficitNs / 1000000;
-            long sleepNs = sleepDeficitNs - sleepMs * 1000000;
-
             long sleepStartNs = System.nanoTime();
+            long currentTimeNs = sleepStartNs;
             try {
-                Thread.sleep(sleepMs, (int) sleepNs);
+                synchronized (this) {
+                    long elapsed = currentTimeNs - sleepStartNs;
```

Comment:
No, it was not correct. Good catch! I've fixed it and validated correctness by running ProducerPerformance manually.